packages/amplify_authenticator/lib/src/widgets/form_field.dart

```diff
@@ -95,64 +106,47 @@ abstract class AuthenticatorFormField<FieldType,
     properties.add(EnumProperty('hintTextKey', hintTextKey));
     properties.add(StringProperty('title', title));
     properties.add(StringProperty('hintText', hintText));
+    properties.add(ObjectFlagProperty<FormFieldValidator<FieldValue>?>.has(
+        'validatorOverride', validatorOverride));
   }
 }
 
-abstract class _AuthenticatorFormFieldState<FieldType,
-        T extends AuthenticatorFormField<FieldType, T>>
+abstract class AuthenticatorFormFieldState<FieldType, FieldValue,
+        T extends AuthenticatorFormField<FieldType, FieldValue, T>>
     extends AuthenticatorComponentState<T> {
   @nonVirtual
   Widget get visibilityToggle => context
       .findAncestorStateOfType<AuthenticatorFormState>()!
       .obscureTextToggle;
 
-  late final List<LoginMechanisms> _loginMechanisms = config.amplifyConfig?.auth
-          ?.awsCognitoAuthPlugin?.auth?['Default']?.loginMechanisms ??
-      const <LoginMechanisms>[];
+  late final List<AwsCognitoUsernameAttributes> _usernameAttributes = config
+          .amplifyConfig
+          ?.auth
+          ?.awsCognitoAuthPlugin
+          ?.auth?['Default']
+          ?.awsCognitoUsernameAttributes ??
+      const <AwsCognitoUsernameAttributes>[];
 
   @nonVirtual
-  TextInputType get usernameKeyboardTypeForAlias {
-    if (_loginMechanisms.contains(LoginMechanisms.preferredUsername)) {
-      return TextInputType.text;
-    } else if (_loginMechanisms.contains(LoginMechanisms.email)) {
-      if (_loginMechanisms.contains(LoginMechanisms.phoneNumber)) {
-        // TODO: can we improve on just a text field
-        // maybe include an icon/toggle to alert user.
-        return TextInputType.text;
-      }
+  TextInputType get usernameKeyboardTypeForUsername {
+    if (_usernameAttributes.contains(AwsCognitoUsernameAttributes.email) &&
+        _usernameAttributes.length == 1) {
       return TextInputType.emailAddress;
-    } else if (_loginMechanisms.contains(LoginMechanisms.phoneNumber)) {
+    } else if (_usernameAttributes
+            .contains(AwsCognitoUsernameAttributes.phoneNumber) &&
+        _usernameAttributes.length == 1) {
       return TextInputType.phone;
     } else {
       // No assumptions made
       return TextInputType.text;
     }
   }
 
-  @nonVirtual
-  ValueChanged<String> get usernameOnChangedForAlias {
-    final List<ValueChanged<String>> ops = [];
-
-    if (_loginMechanisms.contains(LoginMechanisms.preferredUsername)) {
-      ops.add(viewModel.setUsername);
-    }
-    if (_loginMechanisms.contains(LoginMechanisms.email)) {
-      ops.add(viewModel.setEmail);
-    } else if (_loginMechanisms.contains(LoginMechanisms.phoneNumber)) {
-      ops.add(viewModel.setPhoneNumber);
-    }
-    return (String value) {
-      for (var op in ops) {
-        op(value);
-      }
-    };
-  }
-
   /// Callback for when `onChanged` is triggered on the [FormField].
-  ValueChanged<String> get onChanged => (_) {};
+  ValueChanged<FieldValue>? get onChanged => (_) {};
```

Comment:
Doesn't need to be nullable. That's why there is a default value of a no-op.

Reply:
This is due to the fact that each input-specific extension of a field type (ex: `class _SignInTextFieldState extends _SignInFormFieldState`) contains switch cases on the field itself, which include EVERY field. For example, the switch case for `_SignInTextFieldState` will need to handle cases for fields which use the radio button mixin - unless of course we revise this design again. 

Currently, I just return null for those cases, but that's probably not the best way because we have to do `onChanged?.call` elsewhere. Maybe throw some sort of exception?